[nanojsonc] Add new port

saadshams opened the pull request:
<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->

<!-- If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/ -->

<!-- If this PR updates an existing port, please uncomment and fill out this checklist:

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [ ] SHA512s are updated for each updated download
- [ ] The "supports" clause reflects platforms that may be fixed by this new version
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [ ] Any patches that are no longer applied are deleted from the port's directory.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is added to each modified port's versions file.

END OF PORT UPDATE CHECKLIST (delete this line) -->

<!-- If this PR adds a new port, please uncomment and fill out this checklist:

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [ ] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [ ] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [ ] The versioning scheme in `vcpkg.json` matches what upstream says.
- [ ] The license declaration in `vcpkg.json` matches what upstream says.
- [ ] The installed as the "copyright" file matches what upstream says.
- [ ] The source code of the component installed comes from an authoritative source.
- [ ] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is in the new port's versions file.
- [ ] Only one version is added to each modified port's versions file.

END OF NEW PORT CHECKLIST (delete this line) -->

saadshams committed:
Create nanojson (JSON Parser for C/C++)

Event-Driven JSON Parser for C/C++

saadshams committed:
Creating nanojson port

Event-Driven JSON Parser for C/C++

saadshams committed:
init

saadshams committed:
update nanojson

saadshams committed:
Merge branch 'microsoft:master' into nanojson

Neumann-A commented on ports/nanojson/portfile.cmake:
```diff
@@ -0,0 +1,13 @@
+        OUT_SOURCE_PATH source
+        REPO saadshams/nanojson
+        REF 1.0.0
+        SHA512 0
```
```suggestion
        SHA512 d7ac163455b585ac00fafa3cc6c7090ec39e43b45e95f0d7006e56dcedc7bde05070ad339df48b25ff2e3a545fab96ca80eb5f668fc87557add70c3578bb74ba
```

Neumann-A commented on ports/nanojson/vcpkg.json:
```diff
@@ -0,0 +1,9 @@
+  "version-string" : "1.0.0",
+  "description" : "Event-Driven JSON Parser for C/C++",
+  "homepage" : "https://github.com/saadshams/nanojson",
+  "license": "MIT",
```
doesn't agree with https://github.com/mmastrac/nanojson#license

saadshams replied:
There's no comparison; my repo is in C, while the one listed in your comment is in Java. C is way faster than Java.
I have about ~250 LOC compared to the linked repo of 800 LOC.
Further, the repos are not related in any way; the repo under review uses an event- or callback-based approach.

saadshams replied:
ready for review

Neumann-A replied:
Perfect example why it should have a different portname in vcpkg then `nanojson`. Maybe `c-nanojson` if you don't want to apply https://github.com/microsoft/vcpkg/pull/34909/files#r1384511497.

JonLiu1993 commented on ports/nanojson/vcpkg.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "name" : "nanojson",
+  "version-string" : "1.0.0",
```
```suggestion
  "version" : "1.0.0",
```

JonLiu1993 commented on ports/nanojson/vcpkg.json:
```diff
@@ -0,0 +1,9 @@
+  "homepage" : "https://github.com/saadshams/nanojson",
+  "license": "MIT",
+  "maintainers" : [ "Saad Shams" ],
+  "dependencies" : [ ]
+}
```
```suggestion
  "dependencies" : [
      {
      "name": "vcpkg-cmake",
      "host": true
    }
    ]
}

```

JonLiu1993 commented on ports/nanojson/portfile.cmake:
```diff
@@ -0,0 +1,13 @@
+
+vcpkg_cmake_install()
+
+file(INSTALL "${SOURCE_PATH}/COPYING" DESTINATION "${CURRENT_PACKAGES_DIR}/share/nanojson" RENAME copyright)
```
```suggestion
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/COPYING")

```

JonLiu1993 commented on ports/nanojson/COPYING:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
We don't have to provide this.

saadshams replied:
Not sure about this, you mean shall I remove this line?

JonLiu1993 replied:
Yes, we can get it directly from the upstream code via a function. `vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/COPYING")`

saadshams replied:
done.

JonLiu1993 commented:
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

saadshams committed:
update sha512

Co-authored-by: Alexander Neumann <30894796+Neumann-A@users.noreply.github.com>

saadshams committed:
update version

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

saadshams committed:
update copy license

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

saadshams committed:
update dependencies

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

saadshams committed:
update vcpkg

dg0yt commented:
IMO the port should be named  `saadshams-nanojson`. The name is not unique, and this upstream isn't dominant.
https://github.com/search?q=nanojson&type=repositories

The `COPYING` file doesn't belong into the vcpkg repo.

saadshams committed:
update license filename

saadshams commented:
> IMO the port should be named `saadshams-nanojson`. The name is not unique, and this upstream isn't dominant. https://github.com/search?q=nanojson&type=repositories
> 
> The `COPYING` file doesn't belong into the vcpkg repo.

You mean should I delete the COPYING file or LICENSE file?

dg0yt commented:
> You mean should I delete the COPYING file or LICENSE file?

The file `LICENSE` should be removed from this PR to vcpkg.

JonLiu1993 commented on ports/nanojson/vcpkg.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "name" : "nanojson",
```
As @dg0yt said, the port should be named saadshams-nanojson. The name is not unique, and this upstream isn't dominant.

saadshams replied:
I can't rename the port to use my personal name, there are no dominant ports to begin with, and there are no other ports qualified and accepted into json.org. There could be 100 ports out there with the same name, but none of them are listed on the site. Check https://www.json.org/json-en.html under C for the listing. This port will be the dominant one.

saadshams replied:
LICENSE removed.

dg0yt replied:
> I can't rename the port to use my personal name.

It is suggested to rename the port, not the repo, to `<GH organization>-<repository>`. This is feasible and a common request for very young packages.
The vcpkg naming reference is repology.org. I do understand the relevance of json.org, but I have no idea how this list is curated.
Disclaimer: I'm only a community member.

dg0yt replied:
There is a link on that page. Fine. But the screenshot doesn't add any new information.

saadshams replied:
> It is suggested to rename the port, not the repo, to `<GH organization>-<repository>`. This is feasible and a common request for very young packages. The vcpkg naming reference is repology.org. I do understand the relevance of json.org, but I have no idea how this list is curated. Disclaimer: I'm only a community member.

**Historical Significance**: JSON.org is indeed a recognized authority over the JSON format, and by accepting the project "nanoJSON," it pays homage to its shortest codebase and quality standards with thorough unit testing and examples. The name "nanoJSON" from JSON.org can be seen as a nod to the repository's compactness, importance, quality, and ubiquity, showing that your project is built upon a well-established foundation, and vcpkg should respect that. JSON.org judges that from the API, unit tests, and the examples that are also part of the unit tests provided.

**Innovation and Uniqueness**: The project introduces innovations like an event-based API, something that has not been done before. The name "nanoJSON" could signify that it's a compact and efficient solution for handling JSON data, highlighting its uniqueness and forward-thinking approach.

**Project Maturity**: The age of a project is not the sole indicator of its quality or reliability. What matters is the project's ability to fulfill its promises and pass all required tests support JSON format. "nanoJSON" might be a mature and robust project, even if it is relatively new, as evident from the examples and the fact that it fulfills its intended purpose.

**Author's Credibility**: It's important to consider the credibility of the author. The author has a track record of delivering open-source projects successfully and fulfilling promises; this is a strong argument for keeping the name "nanoJSON." The author's reputation and commitment to the project's reliability should be more important than the project's age.

**Recognition of Compatibility**: By aligning with the term "nanoJSON" in the name, you make it clear that "nanoJSON" is intended to be the shortest implementation, fostering trust among developers who are familiar with working in constrained environments. 

If there had been any other previous nanoJSON projects, they would have come forward by this point to apply for the name, let alone receive approval from the JSON.org authority. I'm the first one to do that, irrespective of age.

Neumann-A replied:
> In summary,

Yawn. Let me review that https://github.com/saadshams/nanojson/blob/main/CMakeLists.txt

- Not using `include(CTest)` instead raw `enable_testing()`
- Not making tests/example optional.
- Not defining installing/exporting targets. 
- Not installing usage via nanoJSONConfig.cmake (see https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html). 
- using `CMAKE_C_STANDARD` instead of `target_compile_features( ..... c_std_11)`

- no sanitizer runs in CI.
- no werror ci run

- bad naming of functions. Very likely to conflict. parse_array/parse_object in C code really? That should be at least json_parse_array or better nanojson_parse_error
- no extern "C" to guard for being included in C++
- Why `const char *parentKey` and not `const char * const parentKey` ? (I mean I see the reason in the implementation but from an API point I don't see why parentKey should be changed inside the implementation?)

saadshams replied:
I appreciate the scrutiny. 
Attempted with all except nanojSONConfig.cmake. 
Please review the installing/exporting as I'm new to integration with vcpkg. 

- `include(CTest)` done
- Not making tests/example optional. done
- not defining installing/exporting targets
- set as target_compile_features(nanojson PRIVATE c_std_11)
- no sanitizer runs in CI. added
- no werror ci run. added
- bad naming of functions. added nanojson prefix
- no extern "C". I realize the scope was limited to C (removing C++ support). For C++ there will be another compatible class-based library that I'll publish soon.
- const char *parentKey and not const char * const parentKey: The caller has the option to provide NULL as the parentKey during the initial function call. The function, in turn, examines this value, assigns an empty string if it is NULL, and proceeds to concatenate it with subkeys as it encounters nested objects or arrays. This process ensures the preservation of hierarchical information, which is then passed on to the callback function.

saadshams replied:
@Neumann-A  Added installing usage via nanoJSONConfig.cmake
As mentioned earlier, I'm new with integration with vcpkg, I'd highly appreciate any inputs if I'm missing something.

Ready for review.

saadshams committed:
deleted license

saadshams commented:
Another sign of quality is the passing badge for the CI/CD Pipeline, with flying colors, notably Windows.

Platforms include:
1. Windows
2. Ubuntu (gcc)
3. Ubuntu (clang)

See the badge. "CMake on multiple platforms passing" on the readme.

Please review.

JonLiu1993 commented:
None of the ci tests have been successful so far.
````
CMake Error at /mnt/vcpkg-ci/installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:30 (message):
  SOURCE_PATH must be set
Call Stack (most recent call first):
  ports/nanojson/portfile.cmake:9 (vcpkg_cmake_configure)
  scripts/ports.cmake:168 (include)

````

saadshams commented:
> None of the ci tests have been successful so far.
> 
> ```
> CMake Error at /mnt/vcpkg-ci/installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:30 (message):
>   SOURCE_PATH must be set
> Call Stack (most recent call first):
>   ports/nanojson/portfile.cmake:9 (vcpkg_cmake_configure)
>   scripts/ports.cmake:168 (include)
> ```

@JonLiu1993  Thanks for your response.

1. I have re-pushed the tags 1.0.0 just in case.
2. Secondly, since I'm new to VCPKG, I appreciate your earlier help with refactoring portfile.cmake and vcpkg.json for me. Can you help me with the portfile.cmake here again?

In this case, the error is at line 9 and the complainant says SOURCE_PATH is not set. I'm not sure if we even need this line.
`vcpkg_cmake_configure(SOURCE_PATH ${SOURCE_PATH})`

What really needs to happen is the execution of CMakeLists.txt in the root of my repository.

3. Third, my master branch is called main; do we need to change HEAD_REF from master to main?

Please refactor or advise.

dg0yt commented on ports/nanojson/portfile.cmake:
```diff
@@ -0,0 +1,13 @@
+vcpkg_from_github(
+        OUT_SOURCE_PATH source
```
You put the source path into the variable `source`...

saadshams replied:
Thanks, it has been changed.

saadshams replied:
Can you please guide on how to build locally?

dg0yt commented on ports/nanojson/portfile.cmake:
```diff
@@ -0,0 +1,13 @@
+        HEAD_REF master
+)
+
+vcpkg_cmake_configure(SOURCE_PATH ${SOURCE_PATH})
```
... so use that variable here, not the empty SOURCE_PATH:
```suggestion
vcpkg_cmake_configure(SOURCE_PATH "${source}")
```

saadshams replied:
thanks updated.

dg0yt commented on ports/nanojson/portfile.cmake:
```diff
@@ -0,0 +1,13 @@
+
+vcpkg_cmake_install()
+
+vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
```
and here:
```suggestion
vcpkg_install_copyright(FILE_LIST "${source}/LICENSE")
```

saadshams replied:
how about changing `HEAD_REF master` to `HEAD_REF main`, is it required?

saadshams replied:
please review.

saadshams replied:
```
-- Configuring x64-osx
-- Building x64-osx-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: /usr/local/Cellar/cmake/3.27.7/bin/cmake --build . --config Debug --target install -- -v -j13
```

**install-x64-osx-dbg-out.log**
ninja: error: unknown target 'install', did you mean 'test/all'?

/install-x64-osx-dbg-err.log
Change Dir: '/Users/saad/Documents/vcpkg/buildtrees/nanojson/x64-osx-dbg'

Run Build Command(s): /usr/local/bin/ninja -v -v -j13 install

dg0yt reviewed (commented):
At least try to build locally first, or leave the PR in draft mode.

saadshams committed:
update

saadshams committed:
update hash

saadshams committed:
update

saadshams committed:
update

JonLiu1993 commented:
@saadshams, I will help you fix the ci error.

saadshams commented:
> @saadshams, I will help you fix the ci error.



> @saadshams, I will help you fix the ci error.

@JonLiu1993 FYI, the package is being tested and installed locally. I am looking forward to a review of the following for any edge cases:

**Repo:**

1. CMakeLists.txt
2. cmake/nanoJSONConfig.cmake.in
3. CTestConfig.cmake: CTest and CDash Integration
4. vcpkg.json

**VCPKG repo:**

1. portfile.cmake
2. vcpkg.json

dg0yt commented:
The task of vcpkg maintainers should be focused on the vcpkg port (B), plus basic usage (related to C). A.3 is out of scope.
I repeat the suggestion that the port should be named  `saadshams-nanojson`.

saadshams commented:
> The task of vcpkg maintainers should be focused on the vcpkg port (B), plus basic usage (related to C). A.3 is out of scope. I repeat the suggestion that the port should be named `saadshams-nanojson`.

Sure, I'd appreciate any support you can offer; we can skip A3. @JonLiu1993 When would you be able to help? I assume and hope that with the current state and the progress I have made, it shouldn't take longer than 10–15 minutes of your time to resolve.

I have several other tools and utilities in the pipeline to implement a microservice REST framework in C and I'm waiting on a solution for nanojson to be reused for the others.

Regarding the name change, the port will stay as nanojson, just like other ports. I'm afraid I can't change it.

saadshams committed:
update

saadshams commented:
@JonLiu1993 While I'm discussing the naming with @dg0yt you haven't done the review yet. 

**At least it should be able to install locally in a platform agnostic way** 

Please review VCPKG and the User of the Library project; I'm referencing Static Library as well for perspective.


## Static Library (to give perspective)

**CMakeLists.txt**

```cmake
cmake_minimum_required(VERSION 3.26)
project(nanojson C)

include(GNUInstallDirs) # Include GNUInstallDirs to set installation directories

set(CMAKE_C_STANDARD 11)

add_library(nanojson src/object.c src/array.c)
target_compile_features(nanojson PRIVATE c_std_11)

#target_include_directories(nanojson PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
target_include_directories(nanojson PUBLIC
        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
        $<INSTALL_INTERFACE:include>)

# Installation rules
install(
        TARGETS nanojson
        EXPORT nanojsonTargets
        ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
        INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
)

install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/include/ # installs header
        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/nanojson)

#Exporting Targets
install(EXPORT nanojsonTargets
        FILE nanojsonTargets.cmake
        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/nanojson)

include(CMakePackageConfigHelpers)
configure_file(cmake/nanoJSONConfig.cmake.in ${CMAKE_CURRENT_BINARY_DIR}/nanoJSONConfig.cmake @ONLY)
install(FILES ${CMAKE_CURRENT_BINARY_DIR}/nanoJSONConfig.cmake DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/nanojson)
```

**cmake/nanoJSONConfig.cmake.in**

```cmake
include(CMakeFindDependencyMacro)

# Declare the dependency on nanojson
find_dependency(nanojson REQUIRED)

# Create imported target
add_library(nanojson::nanojson INTERFACE IMPORTED)

# Set target properties
set_target_properties(nanojson::nanojson PROPERTIES
        INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
        INTERFACE_LINK_LIBRARIES "${_IMPORT_PREFIX}/lib/libnanojson.a"
)
```



## VCPKG

**portfile.cmake**

```cmake
vcpkg_from_github(
        OUT_SOURCE_PATH SOURCE_PATH
        REPO saadshams/nanojson
        REF 1.0.0
        SHA512 beb1017eb5242f98d49d39421da92dce6ca4f93d831223070ba48fd69375a6289b281226f14a148945c431b825dbb71c08fc0c288570dcdaa547af3531864d54
        HEAD_REF main
)

vcpkg_cmake_configure(SOURCE_PATH ${SOURCE_PATH})
vcpkg_cmake_install()
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
```

**vcpkg.json**

```json
{
  "name" : "nanojson",
  "version" : "1.0.0",
  "description" : "Event-Driven JSON Parser for C/C++",
  "homepage" : "https://github.com/saadshams/nanojson",
  "license": "BSD-3-Clause",
  "maintainers" : [ "Saad Shams" ],
  "dependencies" : [
    {
      "name": "vcpkg-cmake",
      "host": true
    }
  ]
}
```


## User of the Library 

**CmakeLists.txt**

```cmake
cmake_minimum_required(VERSION 3.26)
project(usagenanojson C)
set(CMAKE_C_STANDARD 11)

find_package(nanojson CONFIG REQUIRED)
add_executable(usagenanojson main.c)
target_link_libraries(usagenanojson PRIVATE nanojson::nanojson)
```

How to find and link libraries in a platform agnostic way. 
I see bunch of files under x4-osx (include/nanojson/parse.h and lib/cmake/nanojson) in UserOfTheLibrary project.

Commit:
[saadshams-nanojson] Add new port

Commit:
update version

saadshams committed:
never agreed to the name, it was published without consent

Osyotr commented:
@saadshams when you create a pull request there's a checkbox that when checked allows maintainers to push changes to your branch. @JonLiu1993 pushed some changes that were required to 
![image](https://github.com/microsoft/vcpkg/assets/8740768/4fde0a4e-54d8-47bd-aa71-de7cd139ae65)
This PR has never been merged to vcpkg, so there's nothing to delete.

saadshams commented:
@Osyotr Thanks for clarification. the problem is that they want to save the name nanojson for someone else who has not even applied to be part of vcpkg distribution channel. There are many other repos but this nanojson is credible and listed at json.org (under C).

Your maintainers are fixated and forcing me to prefix my name to the library names. I'm the first one to come and apply for it but they are denying the name. The question is, who are they saving the name nanojson for?

I have other tools in the pipeline with the same naming convention and providing an ecosystem for lightweight microservices in C.

1. nanoREST
2. nanoCollection
3. nanoMVC framework (based on GoF Design Patterns, minimalistic but fully functional)

JonLiu1993 commented:
@saadshams, Sorry for confusing you, my commit just helps you complete the CI bug fix and the modification to the vcpkg port naming policy.
If you don't agree, I can revert to your original name or you can delete this PR, but I need to explain:
The prefix of the library is just so that users can find the library accurately when using nanojson in vcpkg, because this name is indeed used by many ports. If there are other ports with the same name and want to add new ports to vcpkg, we will not use them. The name is nanojson, but Axx-nanojson or Bxx-nanojson. Each library still obtains the resources corresponding to its own warehouse, and it is easy to distinguish in vcpkg.

![image](https://github.com/microsoft/vcpkg/assets/63675417/550b7036-152a-4422-b42d-9ba15d262dcd)

JonLiu1993 commented:
This PR will not be merged by vcpkg until I add the `reviewed` tag, so there is no need to worry.

saadshams commented:
@JonLiu1993 Thanks for the insights. 

1. I'm going to change gears and name it nanojson-c; there are no conflicts. The library is meant to be for C only.
2. Do you have any HelloWorld demos to demonstrate library creation, the CI/CD of VCPKG and the consumer of the library?

Just for the sake of argument, from the screenshot you shared, the [first entry](https://github.com/mmastrac/nanojson) was in Java , whereas the VCPKG is for C/C++.
The second entry is indeed for C++, with about [2,000 lines of code](https://github.com/ttsuki/nanojson/blob/v3/nanojson3.h), and is still called nanoJSON. For perspective, mine is ~250 LOC. I guess its not much for credibility but

saadshams commented:
@JonLiu1993 I have to mention an exception to the namespace criteria. In the future, there will be a pull request for a library with a name that belongs to an open-source organization for which I'm one of the authors and maintainers, with ports in more than 12 programming languages, using proprietary distribution channels provided by each ecosystem.

The search will show similar names but in languages other than C or C++, using their own respective distribution channels, implying they will never come to vcpkg.

I've authored and finalized C and C++ ports for distribution after I'm done with publishing these dependencies. It's just a heads-up in case the issue of the pre-existence of the name is raised. The name of the library is exclusive, as are the languages (C and C++) and their distribution through vcpkg, so it can't be namespaced, and we will, of course, discuss it in their respective pull requests.

JonLiu1993 commented:
@saadshams, Thanks for your understanding. I think that through this time you will have some understanding of adding new ports in vcpkg and vcpkg related content. I will try my best to help you successfully add your port in vcpkg.

saadshams committed:
rebranded as nanojsonc

saadshams commented:
@JonLiu1993 I have rebranded the project as nanojsonc and I think it perfectly describes its core. I have added a github repo under the same name and updated portfile.cmake and vcpkg.json for your review.

> I will try my best to help you successfully add your port to vcpkg.

I appreciate your help with setting this up since I'm new to it. What I really need is a "pattern" of all industry standard CMake configuration files (a simplest-purest configuration) for the library, vcpkg, and consumer that I can just reuse for a bunch of my other libraries.

The libraries are meant to be really simple, i.e., do one thing and do it really well; at the end, it's just a couple of source files packaged in a bundle, easily deployable, and linkable to a project in a platform-agnostic way, really making it easy for the end-user to simply plug-and-play and feel right at home without any complicated setup, and this is where I'll need your help.

JonLiu1993 commented on ports/nanojsonc/vcpkg.json:
```diff
@@ -0,0 +1,18 @@
+  "version": "1.0.0",
+  "maintainers": ["Saad Shams"],
+  "description": "Event-Driven JSON Parser for C",
+  "homepage": "https://github.com/saadshams/nanojsonc",
```
This link is not available,

saadshams replied:
Sorry my bad, it's public now.

saadshams replied:
I just pushed the tags as well. 1.0.0

JonLiu1993 commented:
@data-queue, Regarding the issue of port naming, could you please take a look?

saadshams commented:
> @data-queue, Regarding the issue of port naming, could you please take a look?

@JonLiu1993 Can you please proceed and publish the package? I have been waiting on this for quite some time now.

JonLiu1993 commented on ports/nanojsonc/portfile.cmake:
```diff
@@ -0,0 +1,22 @@
+        OUT_SOURCE_PATH SOURCE_PATH
+        REPO saadshams/nanojsonc
+        REF "${VERSION}"
+        SHA512 6c1b4c5b43f04b01c94983e89ff6ebcc78cfe8ce060a89a7e2afb8059e50ecd6f3e345189061370f3af7a4cead15da9a29dc8f10f95c88bbce8d195b76936742
```
The hash is incorrect, please install on your local machine and it will tell you the correct hash

saadshams replied:
@JonLiu1993 updated hash.

saadshams committed:
update hash

JonLiu1993 commented:
@saadshams, could you please clarify why nanojson depends on itself, which would cause the error:
````
Maximum recursion depth of 100 exceeded
````
https://github.com/saadshams/nanojsonc/blob/f19f9e5e63b40dfbf33ac0ed279cdb4c570ee846/cmake/nanoJSONcConfig.cmake.in#L4C1-L4C36

saadshams commented:
> @saadshams, could you please clarify why nanojson depends on itself, which would cause the error:
> 

I'm not sure why there's a recursion to begin with; the library is self-sufficient and shouldn't depend on itself. Can you please look at and help with my configuration? Which line to add or remove to make it go through?

**Publisher**: 
Library: [CMakeLists.txt](https://github.com/saadshams/nanojsonc/blob/f19f9e5e63b40dfbf33ac0ed279cdb4c570ee846/CMakeLists.txt)

```
target_include_directories(nanojsonc PUBLIC
        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
        $<INSTALL_INTERFACE:include>)

# Installation rules
install(
        TARGETS nanojsonc
        EXPORT nanojsoncTargets
        ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
        INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
)

# installs header
install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/include/
        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/nanojsonc)

include(CMakePackageConfigHelpers)
configure_file(cmake/nanoJSONcConfig.cmake.in ${CMAKE_CURRENT_BINARY_DIR}/nanoJSONcConfig.cmake @ONLY)
install(FILES ${CMAKE_CURRENT_BINARY_DIR}/nanoJSONcConfig.cmake DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/nanojsonc)
```

**Library:** [cmake/nanoJSONcConfig.cmake.in](https://github.com/saadshams/nanojsonc/blob/f19f9e5e63b40dfbf33ac0ed279cdb4c570ee846/cmake/nanoJSONcConfig.cmake.in)
```
include(CMakeFindDependencyMacro)

# Declare the dependency on nanojsonc
find_dependency(nanojsonc REQUIRED)

# Create imported target
add_library(nanojsonc::nanojsonc INTERFACE IMPORTED)

# Set target properties
set_target_properties(nanojsonc::nanojsonc PROPERTIES
        INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
        INTERFACE_LINK_LIBRARIES "${_IMPORT_PREFIX}/lib/libnanojsonc.a"
)
```

saadshams committed:
update

saadshams committed:
update

saadshams commented:
@JonLiu1993 I have pushed some fixes. Please let me know of any issues.

dg0yt commented on ports/nanojsonc/portfile.cmake:
```diff
@@ -0,0 +1,23 @@
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+set(VCPKG_BUILD_TYPE release) # header-only
```
Only one of these line can be right.

saadshams replied:
Updated. kindly review.

saadshams committed:
update

Commit:
format vcpkg.sjon

Commit:
update version

Commit:
update version

JonLiu1993 commented:
Error log:
````
LINK: command "C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1437~1.328\bin\Hostx64\x86\link.exe test\CMakeFiles\test_array.dir\test_array.c.obj /out:test\test_array.exe /implib:test\test_array.lib /pdb:test\test_array.pdb /version:0.0 /machine:X86 /nologo /DEBUG /INCREMENTAL:NO /OPT:REF /OPT:ICF /subsystem:console nanojsonc.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST:EMBED,ID=1" failed (exit code 1181) with the following output:
LINK : fatal error LNK1181: cannot open input file 'nanojsonc.lib'
````

saadshams commented:
> Error log:
> 
> ```
> LINK: command "C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1437~1.328\bin\Hostx64\x86\link.exe test\CMakeFiles\test_array.dir\test_array.c.obj /out:test\test_array.exe /implib:test\test_array.lib /pdb:test\test_array.pdb /version:0.0 /machine:X86 /nologo /DEBUG /INCREMENTAL:NO /OPT:REF /OPT:ICF /subsystem:console nanojsonc.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST:EMBED,ID=1" failed (exit code 1181) with the following output:
> LINK : fatal error LNK1181: cannot open input file 'nanojsonc.lib'
> ```

Not sure why is it can't find nanojsonc.lib, I have followed this [link](https://learn.microsoft.com/en-us/vcpkg/get_started/get-started-packaging?pivots=shell-cmd) 

Can you please guide on what's missing?

dg0yt commented:
> Not sure why is it can't find nanojsonc.lib, I have followed this [link](https://learn.microsoft.com/en-us/vcpkg/get_started/get-started-packaging?pivots=shell-cmd)
> 
> Can you please guide on what's missing?

It means that this lib is not a header-only.

saadshams committed:
update

saadshams commented:
> It means that this lib is not a header-only.

Thanks for pointing that out. Yes it's not a header-only library. 

Updated to use`vcpkg_check_linkage(ONLY_STATIC_LIBRARY)`. Kindly review.

saadshams committed:
update

saadshams commented:
@JonLiu1993 @dg0yt  Please check.

saadshams committed:
update

saadshams committed:
update

saadshams commented:
> @JonLiu1993 @dg0yt Please check.

I'm getting an error on my local installation as i run vcpkg install

CMake Error at /path/to/usagenanojson/vcpkg_installed/x64-osx/share/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake:81 (message):


  'path/to/vcpkg/packages/nanojsonc_x64-osx/debug/share/nanojsonc'
  does not exist.

Call Stack (most recent call first):
  ports/nanojsonc/portfile.cmake:14 (vcpkg_cmake_config_fixup)
  scripts/ports.cmake:168 (include)

Commit:
fix error

Commit:
update version

dg0yt commented:
The upstream cmake config installation is broken. Documentation or a good example would help.
https://github.com/saadshams/nanojsonc/blob/97ae7012e88da1368534f7b6ac162c363000ec6f/CMakeLists.txt#L23-L27

If there is nothing else left for the config to do (components, versions, dependencies), it might be as simple as
~~~diff
 install(EXPORT nanojsoncTargets
-        FILE nanojsoncTargets.cmake
-        NAMESPACE nanojsoncTargets::
-        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/nanojsoncTargets
+        FILE nanojsoncConfig.cmake
+        NAMESPACE nanojsonc::
+        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/nanojsonc
 )
~~~

Assuming the CMake package is `nanojsonc`:
- `nanojsoncTargets` is the "export set" in CMake, an identifier for data handed over between cmake commands.
- There are various options for the config destination, but `Targets` isn't expected to be part of it.
- `Targets` isn't expected to be part of the namespace.
- There must be a `<Pkg>Config.cmake` or `<pkg>-config.cmake` installed to the desired destination.

saadshams committed:
update

saadshams commented:
I have updated the upstream cmake config as per the suggestion.

>>  There must be a <Pkg>Config.cmake or <pkg>-config.cmake installed to the desired destination.

Where do I place this nanojsoncConfig.cmake and what will be the contents of it. For example?

saadshams committed:
update

saadshams committed:
Merge branch 'nanojson' of https://github.com/saadshams/vcpkg into nanojson

saadshams committed:
update

fixed share

saadshams commented:
@dg0yt @JonLiu1993 @Neumann-A 

Gentlemen, Can you please check what's broken?

## library -> CMakeLists

```
# VCPKG
install(TARGETS nanojsonc EXPORT nanojsonc
        ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
        INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
)

configure_package_config_file(build/config.cmake.in 
        ${CMAKE_CURRENT_BINARY_DIR}/nanojsonc-config.cmake
        INSTALL_DESTINATION ${CMAKE_INSTALL_DATADIR}/nanojsonc NO_SET_AND_CHECK_MACRO)

install(FILES
            ${CMAKE_CURRENT_BINARY_DIR}/nanojsonc-config.cmake
            ${CMAKE_CURRENT_BINARY_DIR}/nanojsonc-config-version.cmake
        DESTINATION
            ${CMAKE_INSTALL_DATADIR}/nanojsonc)

install(EXPORT nanojsonc
        FILE nanojsoncConfig.cmake
        NAMESPACE nanojsonc::
        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/nanojsonc
)
```

## build/config.make.in

```
@PACKAGE_INIT@

include(${CMAKE_CURRENT_LIST_DIR}/nanojsonc-targets.cmake)

check_required_components(nanojsonc)
```

saadshams committed:
update

saadshams commented:
@dg0yt @JonLiu1993 @Neumann-A

The library is good to go; please publish it.

**Follow-up question**: What is "git-tree": "3aabebe9cfde27e640f6573331ce668b843b0381", what is the use and where does the value come from? Source: versions/n-nanjosnc.json

saadshams committed:
update

saadshams commented:
@dg0yt @JonLiu1993 @Neumann-A

Hello guys, can you please publish the library. Waiting...

Neumann-A commented:
You dont need to ping me. I have zero decision power here in regards to stuff being merged.

saadshams commented:
> You dont need to ping me. I have zero decision-making power here in regards to stuff being merged.

Thanks for the response. I'm not familiar with the structure and decision process; who do I ping?

JonLiu1993 commented:
@saadshams, sorry for reply you so late, sorry for replying to you so late, I will review this PR carefully today.

Commit:
formating vcpkg.json

Commit:
update version

Commit:
additional empty line

Commit:
update version

JonLiu1993 commented:
Tested usage successfully by `nanojsonc:x64-windows`:
````
Total install time: 3.5 s
The package nanojsonc provides CMake targets:

    find_package(nanojsonc CONFIG REQUIRED)
    target_link_libraries(main PRIVATE nanojsonc::nanojsonc)
````

saadshams commented:
@JonLiu1993 Thanks, no problem.

For information, I wanted to know where the `"git-tree": "xxx"` is coming from in the versions file?

JonLiu1993 commented:
> @JonLiu1993 Thanks, no problem.
> 
> For information, I wanted to know where the `"git-tree": "xxx"` is coming from in the versions file?

You can refer to [this document](https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#versioning) and Ping me if you have any questions.

saadshams commented:
Thanks. In case of version upgrades, I have the following steps outlined; please let me know if I'm missing a step.

1. Update the code on the git and create a new tag on the upstream.
2. Update the ports/nanojsonc/portfile.cmake : REF and SHA512 values
3. Update versions/n-nanjosnc.json: git-tree value

JonLiu1993 commented:
About your second step，these are some steps I summarized as a maintainer. Of course, you as a user don’t need to be so complicated, but they can give you some reference：

1) First, check the points that need to be modified in the vcpkg.json file:

                 1. First check whether the most basic fields in the vcpkg.json file are present, including Source, Version, and Description. In addition, if Homepage is missing, the user needs to be reminded to add it.

                 2. Port update requires setting Port-Version to 0 or deleting it.

                 3. If a port Upstream and we do not support the officially supported triplet, then you need to delete the corresponding triplet in CI.baseline and add it in the CONTROL file

                   Or add supports to the vcpkg.json file, and add an unsupported message to portfile.cmake.

2) Then you need to check the points that need to be modified in the portfile.cmake file:

                 1. Update first needs to pay attention to some deprecated functions (include(vcpkg_common_functions/vcpkg_configure_cmake/vcpkg_install_cmake/...), as well as some requirements

                   New functions used (vcpkg_replace_string/...), or some special functions vcpkg_check_features that need attention (add FEATURES field)

                 2. You also need to pay attention to some common functions vcpkg_fixup_pkgconfig() that may be missing (you need to add this function if the .PC file is generated but the path is wrong)

                 3. There are also some formatting points that need to be paid attention to. For example, the directory behind the port needs to be written as port/${PORT}, and some modification methods that are obviously incorrect in the format need to be pointed out for the user to modify.

                 4. The use of version, version-server, version-date and version-string fields is easy to confuse.

                 5. Generally, we only add a "version" field every time we update the port, so we need to pay attention to whether the "version" field in versions/x-/xxx.json is added correctly.

3) The last thing you need to pay attention to is the update where the patch has changed.

                 1. If the user's pr update patch is deleted, you need to go to the source to see if the patch has been applied to the sourcr code.

                 2. If the user adds a new patch in this update and the modified source content is large, the user needs to submit an issue to Upstream to explain the situation.

                 3. If the user has modified some content in the patch without comments, you need to ask the user.

4) In addition to the content, you also need to check whether the PR title is written in compliance with the specifications.

saadshams commented:
@JonLiu1993 Thanks for the notes, much appreciated.
